File: leetcode/1.java
public class Solution {
    public int[] twoSum(int[] nums, int target) {
        int[] res = new int[2];
        for (int i=0;i<nums.length-1;++i)
        for (int j=i+1;j<nums.length;++j)
        if (nums[i] + nums[j] == target){
            res[0] = i;
            res[1] = j;
            return res;
        }
        return null;
    }
}

//-------------------

public class Solution {
    public int[] twoSum(int[] nums, int target) {
        int[] res = new int[2];
        HashMap<Integer, Integer> h = new HashMap<>();
        for (int i=0;i<nums.length;++i){
            h.put(nums[i], i);
        }
        for (int i=0;i<nums.length;++i){
            if (h.containsKey(target-nums[i])){
                if (h.get(target-nums[i]) == i) continue;
                res[0] = h.get(target-nums[i]);
                res[1] = i;
                return res;
            }
        }
        return null;
    }
}

File: leetcode/1.rs
/// link: https://leetcode.com/problems/two-sum
/// problem: 数组中有两个元素相加等于target，返回其下标
/// solution: 暴力扫
/// solution-fix: 用map存取已扫描过的值

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        for i in 0..nums.len() {
            for j in i + 1..nums.len() {
                if nums[i] + nums[j] == target {
                    return vec![i as i32, j as i32];
                }
            }
        }
        vec![]
    }
}

// ---
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        use std::collections::HashMap;
        let mut m: HashMap<i32, i32> = HashMap::new();
        for i in 0..nums.len() {
            if m.contains_key(&(target - nums[i])) {
                return vec![i as i32, m[&(target - nums[i])]];
            }
            m.insert(nums[i], i as i32);
        }
        vec![]
    }
}
